src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/SlidingWindowRateLimiter.cs

```diff
@@ -0,0 +1,415 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// <see cref="RateLimiter"/> implementation that replenishes permit counters periodically instead of via a release mechanism.
+    /// </summary>
+    public sealed class SlidingWindowRateLimiter : ReplenishingRateLimiter
+    {
+        private int _requestCount;
+        private int _queueCount;
+        private int[] _requestsPerSegment;
+        private int _currentSegmentIndex;
+        private long _lastReplenishmentTick;
+        private long? _idleSince;
+        private bool _disposed;
+
+        private readonly Timer? _renewTimer;
+        private readonly SlidingWindowRateLimiterOptions _options;
+        private readonly Deque<RequestRegistration> _queue = new Deque<RequestRegistration>();
+
+        // Use the queue as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _queue;
+
+        private static readonly RateLimitLease SuccessfulLease = new SlidingWindowLease(true, null);
+        private static readonly RateLimitLease FailedLease = new SlidingWindowLease(false, null);
+        private static readonly double TickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;
+
+        /// <inheritdoc />
+        public override TimeSpan? IdleDuration => _idleSince is null ? null : new TimeSpan((long)((Stopwatch.GetTimestamp() - _idleSince) * TickFrequency));
+
+        /// <inheritdoc />
+        public override bool IsAutoReplenishing => _options.AutoReplenishment;
+
+        /// <inheritdoc />
+        public override TimeSpan ReplenishmentPeriod => _options.Window;
+
+        /// <summary>
+        /// Initializes the <see cref="SlidingWindowRateLimiter"/>.
+        /// </summary>
+        /// <param name="options">Options to specify the behavior of the <see cref="SlidingWindowRateLimiter"/>.</param>
+        public SlidingWindowRateLimiter(SlidingWindowRateLimiterOptions options!!)
+        {
+            _options = options;
+            _requestCount = options.PermitLimit;
+            _requestsPerSegment = new int[options.SegmentsPerWindow];
+            TimeSpan _windowSegmentInterval = _options.Window / _options.SegmentsPerWindow;
```

Comment:
Local variables don't have the `_` prefix
```suggestion
            TimeSpan windowSegmentInterval = _options.Window / _options.SegmentsPerWindow;
```